frontend/javascripts/oxalis/view/right-border-tabs/connectome_tab/connectome_view.js

```diff
@@ -0,0 +1,784 @@
+    } else {
+      // Find out which synapses were deleted and which were added
+      ({ onlyA: deletedSynapseIds, onlyB: addedSynapseIds } = diffArrays(
+        prevFilteredSynapseIds,
+        filteredSynapseIds,
+      ));
+    }
+
+    const layerName = segmentationLayer.name;
+    const skeleton = Store.getState().localSegmentationData[layerName].connectomeData.skeleton;
+    if (skeleton == null) return;
+
+    const { trees } = skeleton;
+
+    if (deletedSynapseIds.length > 0) {
+      const treeIdsToDelete = [];
+      deletedSynapseIds.forEach(synapseId =>
+        findTreeByName(trees, getTreeNameForSynapse(synapseId)).map(tree =>
+          treeIdsToDelete.push(tree.treeId),
+        ),
+      );
+      if (treeIdsToDelete.length) {
+        Store.dispatch(deleteConnectomeTreesAction(treeIdsToDelete, layerName));
+      }
+    }
+
+    if (addedSynapseIds.length > 0 && filteredConnectomeData != null) {
+      const { synapses } = filteredConnectomeData;
+      const newTrees: MutableTreeMap = {};
+      for (const synapseId of addedSynapseIds) {
+        newTrees[synapseId] = synapseTreeCreator(synapseId, synapses[synapseId].type);
+        const synapseNode = synapseNodeCreator(synapseId, synapses[synapseId].position);
+        newTrees[synapseId].nodes.mutableSet(synapseId, synapseNode);
+      }
+
+      Store.dispatch(addConnectomeTreesAction(newTrees, layerName));
+    }
+  }
+
+  updateAgglomerateTrees(
+    prevConnectomeData: ?ConnectomeData,
+    prevFilteredConnectomeData: ?ConnectomeData,
+    prevCheckedKeys: Array<string>,
+  ) {
+    const { segmentationLayer } = this.props;
+    const { connectomeData, filteredConnectomeData, checkedKeys } = this.state;
+
+    if (segmentationLayer == null) return;
+
+    let prevFilteredAgglomerateIds: Array<number> = [];
+    let filteredAgglomerateIds: Array<number> = [];
+    let prevUnfilteredAgglomerateIds: Array<number> = [];
+    let unfilteredAgglomerateIds: Array<number> = [];
+    if (prevFilteredConnectomeData != null) {
+      prevFilteredAgglomerateIds = getAgglomerateIdsFromConnectomeData(prevFilteredConnectomeData);
+    }
+    if (filteredConnectomeData != null) {
+      filteredAgglomerateIds = getAgglomerateIdsFromConnectomeData(filteredConnectomeData);
+    }
+    if (prevConnectomeData != null) {
+      prevUnfilteredAgglomerateIds = getAgglomerateIdsFromConnectomeData(prevConnectomeData);
+    }
+    if (connectomeData != null) {
+      unfilteredAgglomerateIds = getAgglomerateIdsFromConnectomeData(connectomeData);
+    }
+
+    const checkedAgglomerateIds = getAgglomerateIdsFromKeys(checkedKeys);
+    const visibleAgglomerateIds = checkedAgglomerateIds.filter(agglomerateId =>
+      filteredAgglomerateIds.includes(agglomerateId),
+    );
+
+    const reset =
+      prevConnectomeData != null &&
+      connectomeData != null &&
+      prevConnectomeData.connectomeFile.mappingName !== connectomeData.connectomeFile.mappingName;
```

Comment:
I'm likely missing the big picture here, but for example, wouldn't this need the following `||` case? `prevConnectomeData.connectomeFile !== connectomeData.connectomeFile`? So, if the connectomeFile changed, the agglomerate trees would also need to be reset? Or is this handled somewhere else? Or do you assume that this would automatically mean that the `mappingName` also changed? But then, the condition could be more loose.

In general, I find the `componentDidUpdate` logic a bit hard to grasp, since the individual methods which can be called from there also do this kind of `reset` logic. Maybe it's easier to read if the entire "what needs to be updated?"-logic is handled within `componentDidUpdate` and parameters, such as `reset`, are then simply passed to the helper methods?

Reply:
> prevConnectomeData.connectomeFile !== connectomeData.connectomeFile? So, if the connectomeFile changed, the agglomerate trees would also need to be reset?

Since the agglomerate trees are loaded from the agglomerate file, they only need to be reset if the mapping changes (which is indicated by the `mappingName`). The connectome file only holds the agglomerate IDs which are then used to look up the agglomerate skeletons in the agglomerate file.

> In general, I find the componentDidUpdate logic a bit hard to grasp, since the individual methods which can be called from there also do this kind of reset logic.

That's good feedback, I'll think about how to make this easier to grasp!